### PR TITLE
[C++] Refactor StatusOr and StringPiece

### PR DESCRIPTION
**Pull request by Yannic**
This change makes `StatusOr` and `StringPiece` more like
`absl::StatusOr` and `{absl,std}::string_view`.

Note that there is more work required before the Abseil types can be
used as drop-in replacement.

Progress on #3688

### CODE REVIEW

File: src/google/protobuf/stubs/stringpiece.h [hunk 1]
```diff
@@ -391,7 +321,7 @@ class PROTOBUF_EXPORT StringPiece {
 // one of the arguments is a literal, the compiler can elide a lot of the
 // following comparisons.
 inline bool operator==(StringPiece x, StringPiece y) {
-  stringpiece_ssize_type len = x.size();
+  StringPiece::difference_type len = x.size();
   if (len != y.size()) {
```
Comment (acozzette) [inline src/google/protobuf/stubs/stringpiece.h:325]: Some of the test runs are showing an error on this line:
```
../../../src/google/protobuf/stubs/stringpiece.h: In function ‘bool google::protobuf::stringpiece_internal::operator==(google::protobuf::stringpiece_internal::StringPiece, google::protobuf::stringpiece_internal::StringPiece)’:
../../../src/google/protobuf/stubs/stringpiece.h:325:11: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
   if (len != y.size()) {
           ^
```
I think perhaps `len` should be a `size_type`.

Comment (Yannic) [reply src/google/protobuf/stubs/stringpiece.h:324]: Changed to `size_type` (although I wasn't able to repro the error locally). PTAL, thanks!

**Comment by acozzette** [comment]
It looks like there are just a few more errors coming from stringpiece.h in no_warning_test, which tries to make sure everything builds with warnings treated as errors:
```
./google/protobuf/stubs/stringpiece.h: In constructor ‘google::protobuf::stringpiece_internal::StringPiece::StringPiece(const char*, google::protobuf::stringpiece_internal::StringPiece::size_type)’:
./google/protobuf/stubs/stringpiece.h:203:19: error: comparison of unsigned expression >= 0 is always true [-Werror=type-limits]
     assert(len >= 0);
                   ^
./google/protobuf/stubs/stringpiece.h: In member function ‘char google::protobuf::stringpiece_internal::StringPiece::operator[](google::protobuf::stringpiece_internal::StringPiece::size_type) const’:
./google/protobuf/stubs/stringpiece.h:216:17: error: comparison of unsigned expression >= 0 is always true [-Werror=type-limits]
     assert(0 <= i);
                 ^
./google/protobuf/stubs/stringpiece.h: In member function ‘void google::protobuf::stringpiece_internal::StringPiece::remove_prefix(google::protobuf::stringpiece_internal::StringPiece::difference_type)’:
./google/protobuf/stubs/stringpiece.h:222:23: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
     assert(length_ >= n);
                       ^
./google/protobuf/stubs/stringpiece.h: In member function ‘void google::protobuf::stringpiece_internal::StringPiece::remove_suffix(google::protobuf::stringpiece_internal::StringPiece::difference_type)’:
./google/protobuf/stubs/stringpiece.h:228:23: error: comparison between signed and unsigned integer expressions [-Werror=sign-compare]
     assert(length_ >= n);
                       ^
```

**Comment by acozzette** [comment]
I see there are still just a couple warnings left relating to some asserts involving size_t:
```
./google/protobuf/stubs/stringpiece.h: In constructor ‘google::protobuf::stringpiece_internal::StringPiece::StringPiece(const char*, google::protobuf::stringpiece_internal::StringPiece::size_type)’:
./google/protobuf/stubs/stringpiece.h:203:19: error: comparison of unsigned expression >= 0 is always true [-Werror=type-limits]
     assert(len >= 0);
                   ^
./google/protobuf/stubs/stringpiece.h: In member function ‘char google::protobuf::stringpiece_internal::StringPiece::operator[](google::protobuf::stringpiece_internal::StringPiece::size_type) const’:
./google/protobuf/stubs/stringpiece.h:216:17: error: comparison of unsigned expression >= 0 is always true [-Werror=type-limits]
     assert(0 <= i);
                 ^
```
It might make sense to just delete those asserts since with a `size_t` they are always true.

**Comment by Yannic** [comment]
Thanks, removed! PTAL

File: src/google/protobuf/stubs/stringpiece_unittest.cc [hunk 2]
```diff
@@ -305,16 +305,9 @@ TEST(StringPiece, STL2) {
   const StringPiece a("abcdefghijklmnopqrstuvwxyz");
   const StringPiece b("abc");
   const StringPiece c("xyz");
-  StringPiece d("foobar");
```
Comment (acozzette) [inline src/google/protobuf/stubs/stringpiece_unittest.cc:308]: What's the rationale for removing `d` and the test assertions that use it?

**Review by acozzette — CHANGES_REQUESTED** [review]

**Comment by acozzette** [comment]
Thanks, @Yannic!

File: src/google/protobuf/stubs/stringpiece.h [hunk 3]
```diff
@@ -154,52 +154,26 @@
...
+  using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+  using reverse_iterator = const_reverse_iterator;
+  using size_type = size_t;
+  using difference_type = std::ptrdiff_t;
+
  private:
   const char* ptr_;
-  stringpiece_ssize_type length_;
-
-  // Prevent overflow in debug mode or fortified mode.
```
Comment (haberman) [inline src/google/protobuf/stubs/stringpiece.h:184]: This PR removed the length checking code, but if we are matching ABSL and STL they both have a `max_size()`. In the case of ABSL it is `std::numeric_limits<difference_type>::max()`: https://github.com/abseil/abseil-cpp/blob/1ae9b71c474628d60eb251a3f62967fe64151bb2/absl/strings/string_view.h#L524-L529

Can we add this checking code back with a max size to match ABSL?

**Pull request merged**